feat: add basic Docker support (#14)
autowarefoundation/autoware: docker/README.md

# Docker images for Autoware

## Prerequisites

- [Docker](https://docs.docker.com/engine/install/ubuntu/)
- [rocker](https://github.com/osrf/rocker)

>   We use `rocker` to enable GUI applications such as `rviz` and `rqt` on Docker Containers.

Author: Actually, it was intended. I'll add `- ` for these lines.

>   We use `rocker` to enable GUI applications such as `rviz` and `rqt` on Docker Containers.

Author: 1ef37665e2832fd3fcbac310c04b3209a85b8fb8

How about this?

>   We use `rocker` to enable GUI applications such as `rviz` and `rqt` on Docker Containers.

Author: I assume okay.
![image](https://user-images.githubusercontent.com/31987104/151932399-f9526bc6-4098-4e55-a87d-0df0a5718764.png)

  Please see [here](http://wiki.ros.org/docker/Tutorials/GUI) for more details.

> The setup script in this repository will install these dependencies.

Author: It's here, it uses Ansible. 
https://github.com/autowarefoundation/autoware/blob/e4631e136b8144a420a8969281bb0c57ad0acf8f/ansible/roles/docker/tasks/main.yaml

I'll write the installation steps here.
https://autowarefoundation.github.io/autoware-documentation/tier4-proposal/installation/autoware/docker-installation/

> The setup script in this repository will install these dependencies.

Reviewer: Okay. Perhaps the line above could be clarified with a link to the installation steps that you will write later? eg:
> The setup script in this repository will install these dependencies, as per the installation steps described [here](https://autowarefoundation.github.io/autoware-documentation/tier4-proposal/installation/autoware/docker-installation/).

Or something like that?

> The setup script in this repository will install these dependencies.

Author: @LalithVipulananthan It's nice! But I wouldn't like to add a reverse reference to autoware-documentation.
So I'd like to write this, how about it?

```markdown
The [setup script](../setup-dev-env.sh) will install these dependencies through the [docker role](../ansible/roles/docker/README.md).
```

> The setup script in this repository will install these dependencies.

Reviewer: That looks fine to me.

## Usage

> We have two types of Docker images: `development` and `prebuilt`.

Author: > Lines 15-22 should go immediately below the title rather than here, as these lines explain what the images do rather than how to use them.

Exactly, thank you!

> We should add a line at the very beginning to say These images provide all of the required dependencies for running Autoware in Docker.

Is this for beginners? I can't imagine some dependencies are missing in a Docker image.

> We have two types of Docker images: `development` and `prebuilt`.

Author: Fixed in 1ef37665e2832fd3fcbac310c04b3209a85b8fb8, but didn't added `These images provide all of the required dependencies for running Autoware in Docker.` yet.

> We have two types of Docker images: `development` and `prebuilt`.

Reviewer: Fair enough. That sentence can be omitted then.

The development image enables you to develop Autoware without setting up the local development environment.

The prebuilt image contains executables and enables you to try out Autoware quickly.
Please note that the prebuilt image is not designed for deployment on a real vehicle!

**Note**: Before pulling these images, please confirm and agree with the [NVIDIA Deep Learning Container license](https://developer.nvidia.com/ngc/nvidia-deep-learning-container-license).

> ### The development image

Author: Could you tell me the reason and nuance, please? :pray:

> ### The development image

Reviewer: I thought I'd replied to this one. I can't really explain it, but here's the OSRF Docker image repo on GitHub. 
https://github.com/osrf/docker_images/blob/master/README.md#ros2--

If you look at the ROS 2 images, they are just listed by their names. So for example, `testing` rather than `the testing image`.

> ### The development image

Author: Hmm, I see. English is very difficult for me. :cry:

> To use the development image image, run the following command:

Author: Okay, will remove it.

```bash
docker run --rm -it -v {path_to_your_workspace}:/autoware ghcr.io/autowarefoundation/autoware-universe:latest
```

> To run with `rocker`:

Reviewer: ```suggestion
- To run with `rocker`:
```

Author: Could you write the reason?

Reviewer: No strong reason. More a kind of stylistic choice to make the text stand out? If you would prefer not to have these lines as bullet points, then feel free to ignore this suggestion.

Author: If there will be many suggestions for the style, I'll change it in another PR.
But will keep it as-is in this PR.

```bash
rocker --nvidia --x11 --user --volume {path_to_your_workspace} -- ghcr.io/autowarefoundation/autoware-universe:latest
```

> If you locate your workspace under your home directory, you can use the `--home` option instead:

Reviewer: ```suggestion
- If you locate your workspace under your home directory, you can use the `--home` option instead:
```

Author: The same as https://github.com/autowarefoundation/autoware/pull/14#discussion_r794627524.

I'd appreciate it if you could write the reasons as well when you point out something. :pray: 
It's for not repeating the same mistakes in the future.

Reviewer: See above. I will try to provide a justification for suggested changes going forwards.

```bash
rocker --nvidia --x11 --user --home -- ghcr.io/autowarefoundation/autoware-universe:latest
```

> To use a customized `.bashrc` for the container:

Reviewer: ```suggestion
- To use a customized `.bashrc` for the container:
```

Author: The same as https://github.com/autowarefoundation/autoware/pull/14#discussion_r794627524.

Reviewer: As noted above, this was purely a stylistic choice that can be ignored if you do not agree with it.

```bash
rocker --nvidia --x11 --user --home --volume $HOME/.bashrc.container:$HOME/.bashrc -- ghcr.io/autowarefoundation/autoware-universe:latest
```

> ### The prebuilt image

Author: 1ef37665e2832fd3fcbac310c04b3209a85b8fb8

To use the prebuilt image, run the following command:

```bash
docker run --rm -it ghcr.io/autowarefoundation/autoware-universe:latest-prebuilt
```

> To run with `rocker`:

Reviewer: ```suggestion
- To run with `rocker`:
```

Author: The same as https://github.com/autowarefoundation/autoware/pull/14#discussion_r794627524.

Author: The same as https://github.com/autowarefoundation/autoware/pull/14#discussion_r794627524.

Reviewer: Stylistic choice that can be ignored if you do not agree with it.

Author: Okay, then let me keep the current style so far. :bow:

```bash
rocker --nvidia --x11 --user --volume {path_to_your_workspace} -- ghcr.io/autowarefoundation/autoware-universe:latest-prebuilt
```

> If you intend to use pre-existing data such as maps or Rosbags, modify the `--volume` options shown below.

Reviewer: ```suggestion
- If you intend to use pre-existing data such as maps or Rosbags, modify the `--volume` options shown below.
```

Author: The same as https://github.com/autowarefoundation/autoware/pull/14#discussion_r794627524.

Reviewer: Stylistic choice that can be ignored if you do not agree with it.

```bash

> rocker --nvidia --x11 --user --volume {path_to_your_workspace} --volume {path_to_your_map_data} --volume {path_to_your_log_data} -- ghcr.io/autowarefoundation/autoware-universe:latest-prebuilt

Reviewer: For any of the long commands that result in a scrollbar, perhaps consider splitting the command across multiple lines. See the [WireGuard docker README](https://github.com/linuxserver/docker-wireguard) for examples.

Author: Okay, I'll try not to overflow.
![image](https://user-images.githubusercontent.com/31987104/151579822-87e624af-4ff8-4033-84aa-8d709cd65bf6.png)

Author: I assume okay.

```

## Tips

### Precautions for not using `rocker`

> If you run the following command, it uses the `root` permission.

Author: Do we need `However` here? I feel it can be removed.

> If you run the following command, it uses the `root` permission.

Author: BTW, I'll replace `permissions` with `privileges`.

> If you run the following command, it uses the `root` permission.

Author: 1ef37665e2832fd3fcbac310c04b3209a85b8fb8

I wrote the following, is it okay?

```
If either image is run without `rocker`, then `root` privileges will be used.
This can affect your local environment.
```

> If you run the following command, it uses the `root` permission.

Reviewer: You can combine those sentences I think.
> If either image is run without `rocker`, then `root` privileges will be used which can affect your local environment.

> If you run the following command, it uses the `root` permission.

Reviewer: This is fine.

```sh-session
$ docker run --rm -it -v {path_to_your_workspace}:/autoware ghcr.io/autowarefoundation/autoware-universe:latest
# colcon build
```

> It means it affects

Author: Ah, I was writing something but it's not completed.

> It means it affects

Author: I changed here to `It affects your local environment.`, is it okay or still should be removed?

> It means it affects

Author: 1ef37665e2832fd3fcbac310c04b3209a85b8fb8

Merged with the previous sentence.

> It means it affects

Author: I assume it's okay seeing this comment.
https://github.com/autowarefoundation/autoware/pull/14#discussion_r796282382

```sh-session
# exit
$ rm build/COLCON_IGNORE
rm: remove write-protected regular empty file 'build/COLCON_IGNORE'? y
rm: cannot remove 'build/COLCON_IGNORE': Permission denied
```

> There are several recommended ways to prevent this.

Reviewer: ```suggestion
There are several ways to prevent this error from occurring:
```

Author: 1ef37665e2832fd3fcbac310c04b3209a85b8fb8

> 1. Prepare a dedicated workspace for the docker image.

Reviewer: If `rocker` is recommended, then shouldn't it be at the top of the list? Or is preparing a dedicated workspace our top recommendation?

Author: Mm, either one is fine. There are no priorities.
I just wanted to place rocker and VSCode side by side.

Reviewer: See https://github.com/autowarefoundation/autoware/pull/14/files#r796285036

Author: @LalithVipulananthan I'm sorry I can't find what you point out in the link (the browser doesn't auto-focus.). Could you attach a screenshot or copy the text?

Author: Ah, this?
![image](https://user-images.githubusercontent.com/31987104/151944514-58429d6a-c9dc-48f1-a630-baf19b29d768.png)

2. Use `rocker`.

> 3. Use `Visual Studio Code Remote - Containers`.

Reviewer: ```suggestion
3. Use Visual Studio Code's `Remote - Containers` extension .
```

Author: 1ef37665e2832fd3fcbac310c04b3209a85b8fb8

>    You can use following settings. It creates a user account in a similar way as `rocker`.  

Reviewer: ```suggestion
To use the Remote-Containers extension, the following settings can be used to create a user account in a similar way to `rocker`.  
```

Author: Applied in 1ef37665e2832fd3fcbac310c04b3209a85b8fb8, but I used `the extension` instead.

Author: I assume this can be resolved.

>    Please see [this document](https://code.visualstudio.com/remote/advancedcontainers/add-nonroot-user) for more details.

Reviewer: ```suggestion
Refer to [this document](https://code.visualstudio.com/remote/advancedcontainers/add-nonroot-user) for more details.
```

Author: Could you tell me the nuance for:
- Removing `Please`.
- Replacing `see` with `refer to`

Also, should I replace this line, too?

```
  - Please see [here](http://wiki.ros.org/docker/Tutorials/GUI) for more details.
```

Reviewer: Stylistic choice again. I was having trouble coming up with a half-decent reason for it, but then I found this English Language Stack Exchange answer that answers it for me. 

In short, `refer` feels like the thing being referenced is an external link or document. `see` fits better when the reference is within the same document.
https://english.stackexchange.com/a/553294

Reviewer: I should also explain why I took out the "please" - my understanding is that for technical writing, that word is often left out when giving instructions or when suggesting that a user look at some other documentation or reference ([here's the Tableau Server documentation to give one example](https://help.tableau.com/current/server/en-us/setup.htm))

Author: Thank you, I understand!
Regarding `please`, I'll remove them in the document files, and leave them if they're in the interactive console.

Author: 78cf53d377430ae9a7c89db2c46a7664c4aa8f39

   ```json
   // .devcontainer/devcontainer.json
   {
     "name": "Autoware",
     "build": {
       "dockerfile": "Dockerfile"
     },
     "remoteUser": "autoware",
     "settings": {
       "terminal.integrated.defaultProfile.linux": "bash"
     }
   }
   ```

   ```docker
   # .devcontainer/Dockerfile
   FROM ghcr.io/autowarefoundation/autoware-universe:latest

   ARG USERNAME=autoware
   ARG USER_UID=1000
   ARG USER_GID=$USER_UID

   RUN groupadd --gid $USER_GID $USERNAME \
     && useradd --uid $USER_UID --gid $USER_GID -m $USERNAME \
     && apt-get update \
     && apt-get install -y sudo \
     && echo $USERNAME ALL=\(root\) NOPASSWD:ALL > /etc/sudoers.d/$USERNAME \
     && chmod 0440 /etc/sudoers.d/$USERNAME
   ```

> ### Using other versions of images than `latest`

Reviewer: ```suggestion
### Using versions other than `latest`
```

Author: Mm, I'd like to write `images`...:thinking: Because it might mean the `latest` Docker client.
Are there any alternative sentences?

Reviewer: How about this?
> Using Autoware Docker images other than `latest`

You could also potentially omit `Autoware` from above.

Author: Looks nice! Thank you.

Author: 0ae86b1398968047cafbc817dc27253f9ae34cca

> There are also images versioned based on the `date` or `SemVer`.

Reviewer: ```suggestion
There are also image versions available based on `date` or `SemVer`.
```

What is `SemVar` by the way?

Author: Could you tell me the nuance? (I'm sorry for asking so many times.)

> What is SemVar by the way?

https://semver.org/
But I'll change this to `tag`.

Reviewer: "version" is not a verb so "versioned" looks wrong to me. I have a feeling that it is being used like this by developers however.

> https://semver.org/
> But I'll change this to `tag`.

Okay. In which case, I think it would be better to clarify the meaning of `SemVar` by including a link to the page above, or just putting `(semantic version)` after `SemVar`. Changing `SemVar` to `tag` makes it even less clear what is meant. 

An alternative sentence suggestion:
> There are also other images available that are tagged by either `date` or `SemVer` (semantic version).

Author: Actually, a tag sometimes isn't SemVer because it has a suffix like `v0.1.2-beta`, so writing `SemVer` here is not strictly correct. :thinking:

Author: @LalithVipulananthan How about `release tag` or `software release tag`?

Author: Used `release tag` in 766f9e8a21a12a73aaea7b16e6fc51fcbee7d506.

Please use them when you need a fixed version of the image.  
The list of versions can be found [here](https://github.com/autowarefoundation/autoware/packages).

> ### Building Docker images on your local machine

Reviewer: I think this should be included in its own sub-heading (##) below Usage but above Tips. The very bottom of the page doesn't seem like the right place for this information.

Author: Why do you think so?
I believe users don't use this so often.

Reviewer: If users won't use it very often, then leaving it at the bottom is fine. I still think that building a local image is an alternative use case for the repository, rather than a tip, so it should have its own sub-heading and not be part of the Tips section (this is something I've seen in other GitHub repos containing Docker images)

Author: I feel if we prepare a sub-heading we need a sentence like this:

```markdown
## Building Docker images on your local machine

In case you want to ..., you can build images locally and use them.
``` 

I'm not sure what to write for `...`, so I'd like to keep it as a tip for now.

@LalithVipulananthan Is it okay?

Reviewer: That's fine. We can always revisit at a later date.

Reviewer: The example I was referencing was [found here](https://github.com/linuxserver/docker-wireguard): 
![image](https://user-images.githubusercontent.com/63835446/152083194-bcca61b4-6ac5-41ec-9620-baa180654a4e.png)

If you prefer to keep it as a tip then that is also fine.

Author: Okay, thank you!
I'll add `development purposes` and create a section.

Author: 766f9e8a21a12a73aaea7b16e6fc51fcbee7d506

> If you build the images locally instead of pulling them from the server, run the following command:

Reviewer: ```suggestion
If you want to build the images locally instead of pulling them from the server, run the following command:
```

Author: Could you tell me the nuance?

Reviewer: If my understanding is correct, the whole point of the repo is to enable people to use images that have already been created, and so building the images locally is a deliberate choice and something that someone _wants_ to do versus something that they _need_ to do. It's a minor use case after all, as you pointed out.

If it still looks weird, you could just replace it with:
> To build the images locally, run the following command:

The reason for omitting the "instead of pulling them from the server" is that this is implied by doing a local build. Keeps the sentence shorter and more to the point.

Author: Thank you for your clear answer!
Since I like simple expressions, I'll use `To build the images locally, run the following command:`.

Author: 0ae86b1398968047cafbc817dc27253f9ae34cca

  -- ghcr.io/autowarefoundation/autoware-universe:latest
```

### Prebuilt image

```bash
docker run --rm -it \
  ghcr.io/autowarefoundation/autoware-universe:latest-prebuilt
```

To run with `rocker`:

```bash
rocker --nvidia --x11 --user \
  --volume {path_to_your_workspace} \
  -- ghcr.io/autowarefoundation/autoware-universe:latest-prebuilt
```

If you intend to use pre-existing data such as maps or Rosbags, modify the `--volume` options shown below.

```bash
rocker --nvidia --x11 --user \
  --volume {path_to_your_workspace} \
  --volume {path_to_your_map_data} \
  --volume {path_to_your_log_data} \
  -- ghcr.io/autowarefoundation/autoware-universe:latest-prebuilt
```

## Tips

### Precautions for not using `rocker`

If either image is run without `rocker`, then `root` privileges will be used.
This can affect your local environment.

```sh-session
$ docker run --rm -it -v {path_to_your_workspace}:/autoware ghcr.io/autowarefoundation/autoware-universe:latest
# colcon build
# exit
$ rm build/COLCON_IGNORE
rm: remove write-protected regular empty file 'build/COLCON_IGNORE'? y
rm: cannot remove 'build/COLCON_IGNORE': Permission denied
```

> There are several ways to prevent this error from occurring:

Reviewer: ```suggestion
To prevent this error from occurring, there are two suggested methods:
```

Okay, the reason for this suggestion is:

- The lines above clearly state that this error can occur as a result of _not_ using `rocker` 
- So saying that `rocker` is a method seems like it's over-stating the point a little
- The sentence above is really saying `To prevent this error occurring (without using `rocker`)`

I hope that makes sense. If you really want to keep `rocker` in as a method of preventing the error, then I would put it at the top. However, I will defer to your judgement on this.

Author: Thank you! I've removed `rocker` from the option.
But I want to add `without rocker`.

```
To prevent this error from occurring without `rocker`, there are two suggested methods:
```

Author: 78cf53d377430ae9a7c89db2c46a7664c4aa8f39

Reviewer: Maybe this?
> To prevent this error occurring when `rocker` is not used, there are two suggested methods:

Author: :+1:

Author: Thank you, fixed in e99e6f9abeb02bb0ff503d8f5d11da98c1cae30a.

# Dockerfile for Autoware

## Prerequisites

- [Docker](https://docs.docker.com/engine/install/ubuntu/)
- [rocker](https://github.com/osrf/rocker)

  We use `rocker` to enable GUI applications on Docker Containers.
  Please see [here](http://wiki.ros.org/docker/Tutorials/GUI) for more details.

The setup script in this repository will install these dependencies.

## Usage

Note: Before pulling these images, please confirm and agree with the [NVIDIA Deep Learning Container license](https://developer.nvidia.com/ngc/nvidia-deep-learning-container-license).

### Development image

> This image is for developing Autoware without setting up the local development environment.

Reviewer: If an introductory text is added to explain the purpose of the repository and the difference between the two images, then I think that this line can be removed.

Author: How about this?
https://github.com/autowarefoundation/autoware/pull/14/files#diff-da8fcbe728a9172b578e5d754f8e2df214c658c4321f610e63dd68bea828ab49R15-R20

> You can use this image as below.

Reviewer: ```suggestion
To use the development image, run the following command:
```

Author: Do we need to repeat the word `development` here?
I know it's so because you say so, but let me confirm why it's necessary.

Author: I'll resolve this since you did a new review.

> If you do not mount your workspace by `-v` option, the workspace used for building `prebuilt` image is opened, but it is usually useless.

Author: Yes I know, I'm sorry. It was difficult to explain. :cry: 
I'll remove this.

> If you do not mount your workspace by `-v` option, the workspace used for building `prebuilt` image is opened, but it is usually useless.

Reviewer: No need to apologise! To be honest, if someone decides to omit parts of the command as stated in the usage guide, I do not think it is necessary to explain what may happen as a consequence.

> If you do not mount your workspace by `-v` option, the workspace used for building `prebuilt` image is opened, but it is usually useless.

Author: I'm relieved. :smile:

> If you do not mount your workspace by `-v` option, the workspace used for building `prebuilt` image is opened, but it is usually useless.

Author: a3408bf27422bd592ac2cd80e85a73d91d97a549

```bash
docker run --rm -it -v {path_to_your_workspace}:/autoware ghcr.io/autowarefoundation/autoware-universe:latest
```

> It is useful to use `rocker` if you launch `rviz` or other GUI applications.

Reviewer: To use the development image with `rocker` in order to run `rviz` or any other GUI applications, run the following command:

> Please note that `setup.bash` is not automatically sourced with `rocker`.

Author: I'm sorry I couldn't understand this question well.
What I meant here is, although we source `setup.bash` [here](https://github.com/autowarefoundation/autoware/pull/14/files#diff-36a199a2dbe6455795f2485b956ac1791a8e001492166414332288a216a9aa55R51), it has no effect if you use `rocker`, so you have to manually source it after launching the DOcker image.

```bash
rocker --nvidia --x11 --user --volume {path_to_your_workspace} -- ghcr.io/autowarefoundation/autoware-universe:latest
```

It is also useful to use `--home` option instead of specifying each `--volume`.
Please note that you should not write settings that depend on your local machine, which causes errors.

```bash
rocker --nvidia --x11 --user --home -- ghcr.io/autowarefoundation/autoware-universe:latest
```

### Prebuilt image

This image is for quickly testing Autoware's feature. Please note that it is not designed to deploy on real vehicles.

> You can use this image as below.

Reviewer: ```suggestion
To use the prebuilt image, run the following command:
```

```bash
docker run --rm -it ghcr.io/autowarefoundation/autoware-universe:latest-prebuilt
```

> If you use `rocker`:

Reviewer: ```suggestion
To use the prebuilt image with `rocker`:
```

```bash
rocker --nvidia --x11 --user --volume {path_to_your_workspace} -- ghcr.io/autowarefoundation/autoware-universe:latest-prebuilt
```

> If you use some map data and log data, add `--volume` options as below or use `--home` option and place the data under your home directory.

Reviewer: ```suggestion
If you intend to use pre-existing data such as maps or Rosbags, either modify the `--volume` options shown below or alternatively replace them with the `--home` option and place the required data under your home directory.
```
